### PR TITLE
feat(account-management): report IdP-managed field on update

### PR DESCRIPTION
**Pull request by fluiderson**
## Summary by CodeRabbit

* **New Features**
  * Added clear validation responses when updating identity-provider-managed, non-writable user fields.
  * Identifies all affected fields with HTTP 400 errors and structured `IDP_MANAGED_FIELD` details.
  * Allows updates to other writable fields when managed fields are locked.

* **Documentation**
  * Updated API documentation with managed-field validation behavior and error-response examples.
  * Clarified point lookups using `limit=1`.

* **Bug Fixes**
  * Preserved provider error details while preventing sensitive information from appearing in logs.

### CODE REVIEW

**Comment by fluiderson** [comment]
@coderabbitai review

**Comment by fluiderson** [comment]
@coderabbitai review

**Comment by fluiderson** [comment]
@coderabbitai review

File: gears/system/account-management/account-management/src/api/rest/routes/users.rs [hunk 1]
```diff
@@ -142,7 +155,12 @@ pub(super) fn register_users_routes(mut router: Router, openapi: &dyn OpenApiReg
              body projected through `gts.cf.core.am.user.v1~`. An empty patch is rejected \
              with `code=validation`. Unlike DELETE, a patch against a user the IdP reports \
              absent returns 404 -- it is NOT folded into success. A username rename that \
-             collides with an existing login returns 409.",
+             collides with an existing login returns 409. An attribute the provider \
+             manages and does not allow AM to override (e.g. federated from a read-only \
+             mapper) returns 400 with `context.field_violations[0].field` naming the refused request
```
Comment (diffora) [inline gears/system/account-management/account-management/src/api/rest/routes/users.rs]: **MEDIUM**

The string-literal line ending at `naming the refused request` lost its trailing `\`, so the PATCH description embeds a raw newline plus 13 spaces of indentation — already visible in the committed `docs/api/api.json` (`...refused request\n             property and...`).

The OpenAPI description is a client-facing contract artifact; the stray newline and indent leak into every generated doc and SDK comment.

Append the trailing `\` (matching every sibling line of this literal) and regenerate `docs/api/api.json`.

File: gears/system/account-management/account-management-sdk/src/field.rs [hunk 2]
```diff
@@ -112,6 +125,8 @@ pub enum ValidationReason {
     IdpInvalidInput,
     /// See [`PASSWORD_POLICY`].
     PasswordPolicy,
+    /// See [`IDP_MANAGED_FIELD`].
+    IdpManagedField,
```
Comment (diffora) [inline gears/system/account-management/account-management-sdk/src/field.rs:136]: **MEDIUM**

`ValidationReason` is a growable public SDK enum but is not `#[non_exhaustive]`, so adding `IdpManagedField` is a breaking change for any out-of-repo consumer matching it exhaustively.

The enum already carries the `Unknown(String)` forward-compat catch-all, and every other growable enum in this SDK (`IdpUserOperationFailure`, `IdpUserDuplicateField`, `IdpUserAttribute`) is `#[non_exhaustive]` — this addition and every future reason code repeat the break.

Mark `ValidationReason` as `#[non_exhaustive]` so downstream matches keep compiling via their wildcard arm.

File: gears/system/account-management/account-management-sdk/src/idp_user.rs [hunk 3]
```diff
@@ -875,6 +875,77 @@ pub enum IdpUserOperationFailure {
...
+    /// Distinct from [`Self::UnsupportedOperation`] (the provider does
+    /// not implement `update_user` *at all*, HTTP 501) and from
+    /// [`Self::Rejected`] (the value was bad, not the field). This is a
+    /// per-field capability fact, identical for every caller, so it is
+    /// NOT a permission decision -- AM maps it to the canonical
+    /// validation envelope carrying the structured
+    /// `<field>` / `IDP_MANAGED_FIELD` field-violation tokens (HTTP
+    /// 400) so a client can attribute the refusal to the exact input
+    /// and disable it.
+    FieldNotWritable {
```
Comment (diffora) [inline gears/system/account-management/account-management-sdk/src/idp_user.rs:902]: **MEDIUM**

`FieldNotWritable` carries exactly one `field`, but a merge patch can touch five attributes and a read-only LDAP-federated realm typically locks several at once, so clients discover refusals one round-trip at a time while the canonical `field_violations[]` envelope is already plural.

Widening the variant to a set later is a breaking change for every out-of-repo provider constructing it, and this PR is the release that freezes the SPI shape.

Carry the full refused set (`fields: Vec<IdpUserAttribute>`, documented non-empty) and emit one field violation per entry at the canonical boundary.

File: gears/system/account-management/account-management/tests/api_users_test.rs [hunk 4]
```diff
@@ -670,6 +670,83 @@ async fn update_user_rename_collision_returns_409() {
     assert_eq!(status, StatusCode::CONFLICT);
 }
 
+#[tokio::test]
+async fn update_user_idp_managed_field_returns_400_naming_the_locked_field() {
```
Comment (diffora) [inline gears/system/account-management/account-management/tests/api_users_test.rs:674]: **MEDIUM**

Both locked-attribute REST tests exercise only the set-a-value patch shape; the explicit-clear shape (`{"email": null}`, lowering to `Some(None)`) is documented as equally refused ("scoped to attributes the patch actually touches") but is untested at every level.

The REST-DTO-to-`IdpUserPatch` lowering is exactly where `Some(None)` could collapse to `None` and silently bypass the refusal.

Add a REST test that PATCHes `{"email": null}` against `fake_idp_with_locked_attribute(Email)` and asserts the same 400 with `field == "email"` and `reason == "IDP_MANAGED_FIELD"`.

File: gears/system/account-management/account-management-sdk/src/idp_user.rs [hunk 5]
```diff
@@ -875,6 +875,77 @@ pub enum IdpUserOperationFailure {
...
+
+/// A writable user profile attribute, used to attribute an
+/// [`IdpUserOperationFailure::FieldNotWritable`] refusal to the exact
+/// request field. `#[non_exhaustive]`: the writable surface grows with
+/// [`crate::IdpUserPatch`] without a breaking SDK release.
+///
+/// `password` is deliberately absent: a provider refusing a credential
+/// write reports it through [`IdpUserOperationFailure::PasswordPolicy`],
+/// which already owns the `password` / `PASSWORD_POLICY` tokens.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```
Comment (diffora) [inline gears/system/account-management/account-management-sdk/src/idp_user.rs]: **LOW**

`IdpUserAttribute` is a public fieldless `Copy` token enum but omits `Hash`, and the in-repo fake already documents working around the missing bound with a `Vec` + `contains`.

Providers holding non-writable attribute sets are the primary consumers of this enum; a `HashSet` is their natural shape.

Add `Hash` to the derive list.

File: gears/system/account-management/account-management/src/domain/error.rs [hunk 6]
```diff
@@ -180,6 +180,29 @@ pub enum DomainError {
...
+    /// detail from it, so the wording lives in exactly one place and no
+    /// caller can pair an inconsistent detail/field. Surfaces as HTTP
+    /// 400 `invalid_argument` with `reason = "IDP_MANAGED_FIELD"`.
+    ///
+    /// Deliberately NOT 403: writability is a property of the provider's
+    /// schema, identical for every caller, so no grant would make the
+    /// write succeed. Modelling it as a permission denial would make a
+    /// configuration fact look like a privilege problem.
+    #[error("IdP-managed field not writable ({})", field.as_field_token())]
+    IdpFieldNotWritable {
```
Comment (diffora) [inline gears/system/account-management/account-management/src/domain/error.rs]: **LOW**

`IdpFieldNotWritable` maps to `InvalidArgument` (HTTP 400) but is declared inside the `---- AlreadyExists (HTTP 409) ----` section, following the already-misplaced `IdpPasswordPolicy`.

The enum's section banners are the doc contract reviewers use to eyeball-check the `From<DomainError> for CanonicalError` ladder; a 400 variant filed under 409 defeats that check.

Move `IdpFieldNotWritable` (together with `IdpPasswordPolicy`) under the `---- InvalidArgument (HTTP 400) ----` group, or add a correctly labelled banner above them.

File: gears/system/account-management/account-management/src/infra/sdk_error_mapping_tests.rs [hunk 7]
```diff
@@ -338,6 +338,63 @@ fn idp_password_policy_maps_to_400_with_password_field_violation() {
...
 
+/// An `IdP`-managed attribute reject lands on 400 `invalid_argument`
+/// carrying the *patched property's own name* as
+/// `field_violations[].field` plus the `IDP_MANAGED_FIELD` reason, so a
+/// client can disable exactly that form input. Every
+/// [`account_management_sdk::IdpUserAttribute`] is exercised: the field
+/// token MUST equal the `UserUpdateRequest` JSON property name, because
+/// clients key form-field attribution off that string.
+#[test]
+fn idp_field_not_writable_maps_to_400_with_the_patched_field_violation() {
```
Comment (diffora) [inline gears/system/account-management/account-management/src/infra/sdk_error_mapping_tests.rs:354]: **LOW**

The per-attribute loop asserts `field` and `reason` but never `field_violations[0].description`, and `as_human_phrase` is elsewhere only checked for non-emptiness, so the curated public wording is pinned nowhere.

The variant's doc sells the typed-attribute design as wording that "lives in exactly one place" and cannot drift — without a pinned assertion it drifts silently.

Assert `description` equals the expected curated sentence (e.g. "the email address is managed by the identity provider and cannot be changed through this API") in the loop.

File: gears/system/account-management/account-management/src/api/rest/routes/users.rs (hunk 1 above)
Comment (fluiderson) [reply gears/system/account-management/account-management/src/api/rest/routes/users.rs]: The literal's continuation backslash was restored so the description is one
flowed paragraph again, and the sentence was rewritten for the plural
contract introduced by issue 3 (a refusal can now name several properties):

> Attributes the provider manages and does not allow AM to override (e.g.
> federated from a read-only mapper) return 400 with one
> `context.field_violations[]` entry per refused property, each naming the
> refused request property in `.field` with `.reason=IDP_MANAGED_FIELD`, so a
> client can disable those inputs in one pass; …

`docs/api/api.json` was regenerated via `make openapi` (build the example
server, fetch `/cf/openapi.json`, sort deterministically). The embedded
`\n             ` is gone from the generated artifact.

File: gears/system/account-management/account-management-sdk/src/field.rs (hunk 2 above)
Comment (fluiderson) [reply gears/system/account-management/account-management-sdk/src/field.rs:136]: Added `#[non_exhaustive]` to `ValidationReason`, matching every other growable
enum in the SDK. Also documented *why* both forward-compat mechanisms exist,
since they cover different axes:

> [`Self::Unknown`] covers the same growth on the *value* side (an older SDK
> reading a newer wire string); this attribute covers it on the *type* side.

No in-repo match needed a wildcard arm: the only exhaustive matches
(`from_wire` / `as_wire`) are inside the defining crate, and the
`ValidationReason` referenced across `usage-collector` is a distinct type from
`usage_collector_sdk`.

File: gears/system/account-management/account-management-sdk/src/idp_user.rs (hunk 3 above)
Comment (fluiderson) [reply gears/system/account-management/account-management-sdk/src/idp_user.rs:902]: The SPI variant now carries the full refused set, and the canonical boundary
emits one violation per entry. Changed across the whole chain:

- **SDK** — `IdpUserOperationFailure::FieldNotWritable { fields: Vec<IdpUserAttribute>, detail }`.
  The doc states the set is non-empty, that a provider MUST report every
  refused attribute the patch touched in one failure, and that an
  unattributable refusal belongs in `Rejected` instead.
- **Domain** — `DomainError::IdpFieldNotWritable { fields: Vec<…> }`; the
  `Display` impl joins the field tokens.
- **SPI → domain** (`domain/idp/mod.rs`) — logs the joined field tokens
  un-redacted (they are AM's own tokens, not vendor text) and keeps digesting
  only the provider `detail`.
- **Domain → canonical** (`infra/sdk_error_mapping.rs`) — folds the set into
  the type-state builder, one `<field>` / `IDP_MANAGED_FIELD` violation each.
  The curated wording moved into a single `idp_managed_field_description`
  helper so it still lives in exactly one place across N fields. Duplicates a
  provider may repeat are de-duplicated (order preserved) — cheap now that
  `IdpUserAttribute: Hash` (issue 5).

**Empty-set handling.** `fields` is non-empty *by contract*, but the producer
is out-of-repo, so both boundaries degrade rather than trust it:

- `into_domain_error` maps an empty set to `DomainError::Validation` (the same
  redacted generic rejection as `Rejected`) and logs it as a provider contract
  violation.
- The canonical mapping stays total on its own: a directly-constructed empty
  `IdpFieldNotWritable` falls back to the generic `request` / `VALIDATION`
  pair, because a 400 whose `field_violations[]` is empty attributes the
  refusal to nothing.

New coverage: `field_not_writable_carries_every_refused_attribute` (SDK),
`idp_field_not_writable_emits_one_violation_per_refused_attribute` and
`idp_field_not_writable_with_no_attributes_degrades_to_the_generic_violation`
(canonical mapping), `update_user_idp_managed_fields_are_all_reported_in_one_response`
(REST, below).

File: gears/system/account-management/account-management/tests/api_users_test.rs (hunk 4 above)
Comment (fluiderson) [reply gears/system/account-management/account-management/tests/api_users_test.rs:674]: Added `update_user_idp_managed_field_refuses_an_explicit_null_clear`: PATCHes
`{"email": null}` against `fake_idp_with_locked_attribute(Email)` and asserts
the same 400 / `field == "email"` / `reason == "IDP_MANAGED_FIELD"` triple as
the set-a-value test. The test comment names the specific regression it
guards — the REST-DTO lowering collapsing `Some(None)` to `None` and silently
bypassing the refusal.

Also added `update_user_idp_managed_fields_are_all_reported_in_one_response`
for the plural contract. It patches three attributes against a provider with
three locked ones chosen so the assertion is sharp in both directions:
`email` + `first_name` are touched and locked (refused), `last_name` is locked
but untouched, `display_name` is touched but writable. The expected violation
list is exactly `["email", "first_name"]`.

File: gears/system/account-management/account-management-sdk/src/idp_user.rs (hunk 5 above)
Comment (fluiderson) [reply gears/system/account-management/account-management-sdk/src/idp_user.rs]: Added `Hash` to the derive and documented it as part of the contract ("a
`HashSet<IdpUserAttribute>` is the natural shape on the implementing side").

Followed through on the workaround the fake documented: `FakeIdpPlugin`'s
`locked_attributes` is now a `HashSet` and the stale comment explaining why it
*wasn't* one is gone. Pinned by `user_attribute_is_hashable_for_provider_side_sets`,
so the bound cannot be dropped silently.

File: gears/system/account-management/account-management/src/domain/error.rs (hunk 6 above)
Comment (fluiderson) [reply gears/system/account-management/account-management/src/domain/error.rs]: Moved both `IdpPasswordPolicy` and `IdpFieldNotWritable` out of
`---- AlreadyExists (HTTP 409) ----` and into
`---- InvalidArgument (HTTP 400) ----`, placed after `IdpInvalidInput` so the
three IdP-sourced 400s sit together. The 409 group is now just `AlreadyExists`
+ `UserAlreadyExists`, and the banners once again match the
`From<DomainError> for CanonicalError` ladder. No behaviour change —
`http_status()` already returned 400 for both.

File: gears/system/account-management/account-management/src/infra/sdk_error_mapping_tests.rs (hunk 7 above)
Comment (fluiderson) [reply gears/system/account-management/account-management/src/infra/sdk_error_mapping_tests.rs:354]: The per-attribute loop's table gained a third column with the full expected
sentence, asserted against `field_violations[0].description` — e.g.

> the email address is managed by the identity provider and cannot be changed
> through this API

All five attributes are pinned with literal sentences (not derived from
`as_human_phrase`, which would let the wording drift with the assertion). The
test doc says why the exact string matters rather than just non-emptiness.

**Comment by fluiderson** [comment]
@coderabbitai resume

**Comment by fluiderson** [comment]
@coderabbitai resume

**Comment by fluiderson** [comment]
@coderabbitai resume

**Review by diffora — APPROVED** [review]
